Skip logging rescuable exceptions

islue opened the pull request:
### Summary

ActiveJob always logs exceptions with backtrace from rails 5.2.
For rescuable exceptions, those are defined in rescue_from/retry_on/discard_on, people definitely know what they do since they wrote it in the codes.
Logging well handled exceptions anyway is quite annoying, and it is hard to make a choice to stop log level error for production env.
It doesn't make sense for such a verbose logging, and people can do that in rescue_from if they want, but not vice versa.

islue committed:
Skip logging rescuable exceptions which are defined in rescue_from/retry_on/discard_on.

Update activejob/CHANGELOG.md

Co-authored-by: Rafael França <rafael@franca.dev>

rafaelfranca reviewed (changes requested):
Thank you for the pull request. I agree we need to fix this, but I think this is the wrong fix.

The correct fix would be change the `perform` instrumentation to wrap `perform_now` instead of the `perform` method. This would make sure that the exception is handled before being passed to the instrumentation and it would never log it.

Can you update the PR to do that?

islue committed:
change the `perform` instrumentation to wrap `perform_now` instead of the `perform` method.

islue commented:
Updated according to the correct fix.

* Left `perform_start` in `around_perform` for logging arguments as it is.
* I tried to leave wrapper codes (around_perform replacement) in instrument.rb and logging.rb but failed to find a way. Tell me if you know a better way.

islue commented:
@rafaelfranca any comment or further advice?

rafaelfranca committed:
Refactor the code to keep instrumentation and log inside the modules

Use inheritance to keep the behavior in the right modules.

The order of Instrumentation and Logging had to change to be
flipped to keep the current behavior.

rafaelfranca committed:
Merge remote-tracking branch 'origin/main' into skip-logging-rescuable-exceptions
